docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,199 @@
+Any methods that have the same API across `Series`, `DataFrame`, and `Index` should be defined here.
+Additionally any (internal) methods that could be used to share code between those classes may also be defined here.
+
+The primary internal subclass of `Frame` is `IndexedFrame`, a `Frame` with an index.
+An `IndexedFrame` represents the first type of object mentioned above: indexed tables.
+In particular, `IndexedFrame` is the parent class for `DataFrame` and `Series`.
+Any pandas methods that are defined for those two classes should be defined here.
+
+The second internal subclass of `Frame` is `SingleColumnFrame`.
+As you may surmise, it is a `Frame` with a single column of data.
+This class is the parent for most types of indexes as well as `Series` (note the diamond inheritance pattern here).
+While `IndexedFrame` provides a large amount of functionality, this class is much simpler.
+It adds some simple APIs provided by all 1D pandas objects, and it flattens outputs where needed.
+
+### Indexes
+
+While we've highlighted some exceptional cases of Indexes before, let's start with the base cases here first.
+`BaseIndex` is generally intended to be a true abstract class, i.e. it should contain no implementations.
+Functions may be implemented in `BaseIndex` if they are truly identical for all types of indexes.
+However, currently most such implementations are not applicable to all subclasses and will be eventaully be removed.
+
+Almost all indexes are subclasses of `GenericIndex`, a single-columned index with the class hierarchy:
+`Frame`->`SingleColumnFrame`->`GenericIndex`<-`BaseIndex`.
+Integer, float, or string indexes are all composed of a single column of data.
+Most `GenericIndex` methods are inherited from `Frame`, saving us the trouble of rewriting them.
+
+We now consider the three main exceptions to this model:
+
+- A `RangeIndex` is not backed by a column of data, so it inherits directly from `BaseIndex` alone.
+  Wherever possible, its methods have special implementations designed to avoid materializing columns.
+  Where such an implementation is infeasible, we fall back to converting it to an integer index first instead.
+- A `MultiIndex` is backed by _multiple_ columns of data.
+  Therefore, its inheritance hierarchy looks like `Frame`->``MultiIndex`<-`BaseIndex`.
+  Some of its more `Frame`-like methods may be inherited,
+  but many others must be reimplemented since in many cases a `MultiIndex` is not expected to behave like a `Frame`.
+- Just like in pandas, `Index` itself can never be instantiated.
+  `pandas.Index` is the parent class for indexes,
+  but its constructor returns an appropriate subclass depending on the input data type and shape.
+  Unfortunately, mimicking this behavior requires overriding `__new__`,
+  which in turn makes shared intialization across inheritance trees much more cumbersome to manage.
+  To reenable sharing constructor logic across different index classes,
```

Comment:
I typically write "re-enable" but we can avoid the weirdness by just writing "enable."
```suggestion
  To enable sharing constructor logic across different index classes,
```